collect_app/src/main/java/org/odk/collect/android/formentry/FormEntryViewModel.java

```diff
@@ -305,18 +367,29 @@ public void validate() {
                         error.postValue(new FormError.NonFatal(e.getMessage()));
                     }
 
-                    return result;
-                }, result -> {
-                    isLoading.setValue(false);
-
+                    // JavaRosa moves to the index where the contraint failed
                     if (result instanceof FailedValidationResult) {
-                        refresh();
+                        updateIndex();
                     }
+
+                    return result;
+                }, result -> {
                     validationResult.setValue(new Consumable<>(result));
                 }
         );
     }
 
+    private void preloadSelectChoices() throws RepeatsInFieldListException, FileNotFoundException, XPathSyntaxException {
```

Comment:
> preloadSelectChoices loads the choices into the choices "cache"

Yes, my idea was that this could happen in `loadSelectChoices` so that we don't need that `preloadSelectChoices` which calls `loadSelectChoices`. Something like: https://github.com/getodk/collect/commit/f372fda8b6a3f40e22585e5caca41bc6ef6f342f

Reply:
Ah I see. Hmmm, I'm not sure which one I prefer. I like that that your version isolates the use of the `choices` map (and also reduces total lines of code), but it does feel like we end up with less "command/query separation" - the "preload" is far less explicit now and calling `loadSelectChoices` and not using the result looks a little weird to me. Do you have a strong preference for the one method?

Reply:
> Do you have a strong preference for the one method?

No, it's not a strong preference so let's keep it as it is if you are not sure.